Ensure DDP + Pipe works with find_unused_parameters.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49908 Ensure DDP + Pipe works with find_unused_parameters.**

As described in https://github.com/pytorch/pytorch/issues/49891, DDP +
Pipe doesn't work with find_unused_parameters.

This PR adds a simple fix to enable this functionality. This only currently
works for Pipe within a single host and needs to be re-worked once we support
cross host Pipe.

Differential Revision: [D25719922](https://our.internmc.facebook.com/intern/diff/D25719922/)

pritamdamania committed:
Ensure DDP + Pipe works with find_unused_parameters.

As described in https://github.com/pytorch/pytorch/issues/49891, DDP +
Pipe doesn't work with find_unused_parameters.

This PR adds a simple fix to enable this functionality. This only currently
works for Pipe within a single host and needs to be re-worked once we support
cross host Pipe.

Differential Revision: [D25719922](https://our.internmc.facebook.com/intern/diff/D25719922/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Ensure DDP + Pipe works with find_unused_parameters."


As described in https://github.com/pytorch/pytorch/issues/49891, DDP +
Pipe doesn't work with find_unused_parameters.

This PR adds a simple fix to enable this functionality. This only currently
works for Pipe within a single host and needs to be re-worked once we support
cross host Pipe.

Differential Revision: [D25719922](https://our.internmc.facebook.com/intern/diff/D25719922/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Ensure DDP + Pipe works with find_unused_parameters."


As described in https://github.com/pytorch/pytorch/issues/49891, DDP +
Pipe doesn't work with find_unused_parameters.

This PR adds a simple fix to enable this functionality. This only currently
works for Pipe within a single host and needs to be re-worked once we support
cross host Pipe.

Differential Revision: [D25719922](https://our.internmc.facebook.com/intern/diff/D25719922/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Ensure DDP + Pipe works with find_unused_parameters."


As described in https://github.com/pytorch/pytorch/issues/49891, DDP +
Pipe doesn't work with find_unused_parameters.

This PR adds a simple fix to enable this functionality. This only currently
works for Pipe within a single host and needs to be re-worked once we support
cross host Pipe.

Differential Revision: [D25719922](https://our.internmc.facebook.com/intern/diff/D25719922/)

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -697,7 +701,12 @@ def forward(self, *inputs, **kwargs):
+                if RPC_AVAILABLE and isinstance(self.module, Pipe):
+                    # Unwrap RRef to get real output for Pipe.
+                    # TODO: Needs to be reworked for cross host pipelining.
+                    self.reducer.prepare_for_backward(list(_find_tensors(output.local_value())))
```
Is the RRef always guaranteed to be owned by the node that this runs on? How exactly do we guarantee this?

rohan-varma replied:
Also, is this the only special case that we're going to have where we have to unwrap to get the actual tensors needed in `prepare_for_backwards`? I'm a bit concerned about too tightly coupling Pipe with DDP, when they should probably be as independent as they can be (i.e., ideally DDP won't have to deal with pipe-specific logic). Could we eventually refactor to make that so?

pritamdamania87 replied:
> Is the RRef always guaranteed to be owned by the node that this runs on? How exactly do we guarantee this?

This kinda works right now since Pipe is single host only, but as I mentioned in the comment above, this needs to be reworked for cross host pipelining.

>  I'm a bit concerned about too tightly coupling Pipe with DDP, when they should probably be as independent as they can be (i.e., ideally DDP won't have to deal with pipe-specific logic). Could we eventually refactor to make that so?

This is a good point and when I thought about it a bit more, I think we can make this independent of Pipe for now and also more generic. What we can do is enhance `find_tensors` to unwrap a local RRef and look for tensors inside if it encounters a local RRef.

rohan-varma reviewed (approved):
LGTM, just had a few questions inline

pritamdamania committed:
Update on "Ensure DDP + Pipe works with find_unused_parameters."


As described in https://github.com/pytorch/pytorch/issues/49891, DDP +
Pipe doesn't work with find_unused_parameters.

This PR adds a simple fix to enable this functionality. This only currently
works for Pipe within a single host and needs to be re-worked once we support
cross host Pipe.

Differential Revision: [D25719922](https://our.internmc.facebook.com/intern/diff/D25719922/)

[ghstack-poisoned]

pritamdamania87 commented:
@rohan-varma Updated the PR, could you take another look? Thanks!

rohan-varma reviewed (approved):
Looks good, agreed that enhancing `_find_tensors` is a better approach. It looks like `test_backward_node_failure_python_udf` is failing on this PR, though it's probably flakiness and unrelated.
